includes/AMP/Output_Buffer.php

```diff
@@ -0,0 +1,295 @@
+	}
+
+	/**
+	 * Determine whether output buffering has started.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::start_output_buffering()
+	 * @see Sanitization::finish_output_buffering()
+	 *
+	 * @return bool Whether output buffering has started.
+	 */
+	public function is_output_buffering(): bool {
+		return $this->is_output_buffering;
+	}
+
+	/**
+	 * Finish output buffering.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::start_output_buffering()
+	 *
+	 * @param string $response Buffered Response.
+	 * @return string Processed Response.
+	 */
+	public function finish_output_buffering( string $response ): string {
+		$this->is_output_buffering = false;
+
+		try {
+			$response = $this->prepare_response( $response );
+		} catch ( \Error $error ) { // Only PHP 7+.
+			$response = $this->render_error_page( $error );
+		} catch ( Exception $exception ) {
+			$response = $this->render_error_page( $exception );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Process response to ensure AMP validity.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param string $response HTML document response. By default it expects a complete document.
+	 * @return string AMP document response.
+	 */
+	public function prepare_response( string $response ): string {
+		// Enforce UTF-8 encoding as it is a requirement for AMP.
+		if ( ! headers_sent() ) {
+			header( 'Content-Type: text/html; charset=utf-8' );
+		}
+
+		/**
+		 * Document.
+		 *
+		 * @var Document $dom Document.
+		 */
+		$dom = Document::fromHtml( $response );
+
+		if ( ! $dom instanceof Document ) {
+			return $this->render_error_page( new Exception() );
+		}
+
+		// Make sure scripts from the body get moved to the head.
+		// TODO: Move to its own sanitizer?
+
+		/**
+		 * Scripts in <body>.
+		 *
+		 * @var DOMElement[] $scripts List of scripts.
+		 */
+		$scripts = $dom->xpath->query( '//body//script[ @custom-element or @custom-template or @src = "https://cdn.ampproject.org/v0.js" ]' );
+
+		foreach ( $scripts as $script ) {
+			if ( $script->parentNode ) {
+				$dom->head->appendChild( $script->parentNode->removeChild( $script ) );
+			}
+		}
+
+		$this->sanitize_document( $dom );
+		$this->optimize_document( $dom );
+
+		return $dom->saveHTML();
+	}
+
+
+	/**
+	 * Render error page.
+	 *
+	 * @todo Improve error message.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Throwable $throwable Exception or (as of PHP7) Error.
+	 * @return string Error page.
+	 */
+	private function render_error_page( Throwable $throwable ): string {
+		return esc_html__( 'There was an error generating the web story, probably because of a server misconfiguration. Try contacting your hosting provider or open a new support request.', 'web-stories' );
+	}
+
+	/**
+	 * Sanitizes a document to be valid AMP.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Document $document Document instance.
+	 *
+	 * @return void
+	 */
+	protected function sanitize_document( Document $document ) {
+		$sanitization = new Sanitization();
```

Comment:
Instead of spinning up a new instance, could just inject it? Have a share instance in memory.

Reply:
This is the only place where an instance is needed and created though.

Reply:
We use dependency inject in all other parts of our codebase, why not here? The goals are simple, use classes instances to save memory and allow for easier testing with mocked versions of classes. 

This is one of the few places in the core case, were we spin up a class instance manually and the codebase is not better for it. Why not bring this class inline with other classes?